### PR TITLE
[monodoc] Fixed several issues with `.ToEcmaCref` method.

### PR DESCRIPTION
**Pull request by joelmartinez**
In particular:

- Issues with cref formatting for subtypes.
- Added support for Explicitly implemented members.
- Added support for generic member parameters.

A number of unit tests were added to verify this functionality.

### CODE REVIEW

**Comment by joelmartinez** [comment]
@jonpryor I have several other changes coming up as part of the larger "profiles/filtering" functionality ... these bug fixes were pretty self-contained, and so I've separated them out in an attempt to limit the scope of any given PR only to the minimum required changes :)

File: mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs [hunk 1]
```diff
@@ -454,38 +470,293 @@ public void ExplicitIndexerImplementation ()
...
+		public void ToEcmaCref_Operator ()
+		{
+			var actual = new EcmaDesc {
+				DescKind = EcmaDesc.Kind.Operator,
+				Namespace = "System",
+				TypeName = "Int32",
+				MemberName = "Addition"
+			};
+
+			AssertEcmaString ("O:System.Int32.Addition", actual);
```
Comment (jonpryor) [inline mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs]: While interesting, AFAIK, `O:` isn't a known/existing ECMA prefix.

At least, it's not in [ECMA-334 4th Edition](http://www.ecma-international.org/publications/files/ECMA-ST/Ecma-334.pdf) §E.3.1 **ID string format**.

Is there any reason you can't use `M:` with the IL names, e.g. `M:System.Int32.op_Addition`?

Comment (joelmartinez) [inline mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs]: hmm, interesting ... well, I'm just going off of what's already being used in production, so for example:

http://docs.go-mono.com/?link=O%3aJava.Lang.Float.ExplicitConversion

Seems like this is an instance of where we may have extended the spec long ago

File: mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs [hunk 2]
```diff
@@ -263,9 +255,18 @@ void ConstructCRef (StringBuilder sb)
 			if (DescKind == Kind.Type)
 				return;
 
+			if (ExplicitImplMember != null) {
+				sb.Append ('$');
```
Comment (jonpryor) [inline mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs]: Uh...what?

The nice thing about using ECMA CREFs is that they already exist, there's already tooling for them (`csc /doc`, etc.), and thus there's a general "understanding" of what they represent.

The bad thing about ECMA CREFs is that they already exist, and thus "extending" them isn't necessarily a good idea, as it possibly compromises compatibility with existing toolchains or existing *readers*.

This is the latter: You're apparently using `$` to start the name of an explicit interface name, *but there's already a standard convention for explicit interface implementation*: `s/\//#/g`.

For example, in your build tree you should have `mono/mcs/tools/mdoc/Test/DocTest.xml`, which is the `mcs /doc` output when building DocTest.dll. DocTest.dll contains `MyList<A, B>`, which [explicitly implements `IEnumerable.GetEnumerator()`](https://github.com/mono/mono/blob/master/mcs/tools/mdoc/Test/DocTest-v1.cs#L677).

The resulting `csc /doc` XML?

        <member name="M:Mono.DocTest.Generic.MyList`2.System#Collections#IEnumerable#GetEnumerator">
            <remarks><c>M:Mono.DocTest.MyList`2.System#Collections#GetEnumerator</c>.</remarks></member>

The type, then, is:

    Mono.DocTest.Generic.MyList`2

while the member name is, weirdly-yet-understandably, `System#Collections#IEnumerable#GetEnumerator`.

This "falls out" of the ECMA CREF idea/requirement that for class members, everything from the last `.` to the end of the string is the member name, so a `string.Split()`-like implementation which only splits on `.` will get the correct member name. (This is *also* why constructors in `M` syntax use `#` instead of `.`, e.g. `M:Mono.DocTest.DocAttribute.#ctor(System.String)`.)

In short, using `$` is "weird" and wholly different from what an ECMA CREF should contain, resulting in strings such as `P:System.Web.SessionState.HttpSessionStateContainer$System.Web.SessionState.IHttpSessionState.Item(System.Int32)` which should *really* be `P:System.Web.SessionState.HttpSessionStateContainer.System#Web#SessionState#IHttpSessionState#Item(System.Int32)`.

Comment (joelmartinez) [inline mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs]: Hmm ... well this is another case where I was simply going by long-implemented and public functionality :/ So for example, you can see the cref being generated for this member:

http://docs.go-mono.com/?link=M:System.Collections.Generic.Dictionary<TKey,TValue>$System.Collections.Generic.ICollection<System.Collections.Generic.KeyValuePair<TKey,TValue>>.Add

So I guess the question is ... do we have `ToEcmaCref` conform to the already-implemented functionality in monodoc, or do we change the implementation of the `EcmaHelpSource.GetPublicUrl` to comform to the points you've brought up?

File: mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs [hunk 3]
```diff
@@ -280,6 +281,25 @@ void ConstructCRef (StringBuilder sb)
 			}
 		}
 
+		void AppendGenericArguments (StringBuilder sb, IEnumerable<EcmaDesc> arguments, bool isNumeric, int argumentsCount)
+		{
+			if (arguments != null && isNumeric) {
+				sb.AppendFormat ("`{0}", argumentsCount);
+			} else if (arguments != null) {
+				sb.Append ('<');
```
Comment (jonpryor) [inline mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs]: Given that the provision of type arguments within CREFs is a non-standard extension anyway, perhaps we should just use/standardize on `{}` instead of `<>`? These values are often stored in XML, and using `<>` will require escaping them, while `{}` won't.

File: mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs [hunk 4]
```diff
@@ -454,38 +470,293 @@ public void ExplicitIndexerImplementation ()
...
 		}
 
-		IEnumerable<Node> GetLeaves (Node node)
+		[Test]
+		public void ToEcmaCref_SimpleType ()
 		{
-			if (node == null)
-				yield break;
+			var actual = new EcmaDesc {
+				DescKind = EcmaDesc.Kind.Type,
```
Comment (jonpryor) [inline mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs]: `EcmaDesc.Kind` is presumably a nested type, which should [normally be avoided](https://msdn.microsoft.com/en-us/library/ms229027(v=vs.110).aspx)

> **AVOID** publicly exposed nested types. The only exception to this is if variables of the nested type need to be declared only in rare scenarios such as subclassing or other advanced customization scenarios.

Hopefully that type isn't public...

File: mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs [hunk 5]
```diff
@@ -454,38 +470,293 @@ public void ExplicitIndexerImplementation ()
...
+		}
+
+		[Test]
+		public void ToEcmaCref_NestedType ()
+		{
+			var actual = new EcmaDesc {
+				DescKind = EcmaDesc.Kind.Type,
+				Namespace = "System.IO",
+				TypeName = "Path",
+				NestedType = new EcmaDesc { DescKind = EcmaDesc.Kind.Type, TypeName = "TheNestedType" },
```
Comment (jonpryor) [inline mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs]: You should also test *multiple nested types*, e.g. [`Mono.DocTest.NestedClass.Double.Triple.Quadruple](https://github.com/mono/mono/blob/master/mcs/tools/mdoc/Test/DocTest-v1.cs#L173), which is insanely nested. :-)

File: mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs (hunk 2 above)
Comment (jonpryor) [inline mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs]: This is also why nested types use `+` instead of `.`, to facilitate a simple `string.Split()`-based parsing strategy. Consider:

    var parts = "A.B.C.D.E".Split(".");

For *members* (`F` fields, `M` methods, `P` properties, etc.) the *last* element in the array is the member name, the second-to-last element is the type name -- including nested types -- and everything prior to that is part of the namespace.

For *types*, the last element of the array is the typename, and everything preceding is the namespace.

For *namespaces*, the entire array is the namespace.

This is why all CREFs require a prefix, so you know how to interpret the array. `System.Object` is meaningless; `T:System.Object` can be parsed.

File: mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs (hunk 5 above)
Comment (jonpryor) [inline mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs]: Coding convention: initializers should always span multiple lines:

    var actual = new EcmaDesc {
        // ...
        NestedType = new EcmaDesc {
            DescKind = EcmaDesc.Kind.Type,
            TypeName = "TheNestedType",
        },
    };

File: mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs (hunk 4 above)
Comment (joelmartinez) [inline mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs]: it is public: https://github.com/mono/mono/blob/master/mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs#L13

File: mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs (hunk 2 above)
Comment (jonpryor) [inline mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs]: > do we have ToEcmaCref conform to the already-implemented functionality in monodoc, or do we change the implementation of the `EcmaHelpSource.GetPublicUrl` to comform to the points you've brought up?

I'm inclined to say that we should fix `EcmaHelpSource.GetPublicUrl()`, except for one minor problem: `#` is the anchor for URLs, so a URL such as this has little chance of making sense:

    http://docs.go-mono.com/?link=M:System.Collections.Generic.Dictionary<TKey,TValue>.System#Collections#Generic#ICollection<System.Collections#Generic.KeyValuePair<TKey,TValue>>#Add

To even begin to make that work, you'd have to change the URL scheme to be...something else. (Wasn't there some "newfangled" URL convention a few years ago where everything of interest was after the anchor?)

So I'm torn: for XML/file use, `#` should be used, but for URLs... I'm not sure.

Perhaps replace `#` with `-`?

**Comment by joelmartinez** [comment]
@jonpryor: Perhaps a bit more context would help this conversation. The underlying reason that drove me to make these improvements was to support the `mdoc profile` feature that I'm working on ... as we discussed, this will make a profile from a set of assemblies that will be included in the .zip, and then this information could be used at runtime to filter the tree.

The way I was planning on implementing that was for the profile XML to include CREF urls for every namespace, type, and member ... so that when the API is requested to filter the tree, I can compare the cref in the profile, to the `PublicUrl` given by the node. If they match, it's included, if not ... it's not in the profile and will be filtered out.

For my purposes, I need these values to match ... which is why I tailored my output to what monodoc is currently generating. So the question then becomes:

- Do I change the existing functionality to match the ideal/spec?
- Or, do I continue along this path because it is what monodoc already generates?

I'm happy to go whichever direction you think is best; though as you can probably guess, I'm leaning towards the path of least resistance of simply using the version of the crefs that's already in production :P

File: mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs (hunk 2 above)
Comment (joelmartinez) [inline mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs]: ahh, you know, that's a great point ... `#` is definitely a non-starter as it doesn't even get sent to the web server :/

**Comment by joelmartinez** [comment]
@jonpryor ok, I've made the following changes:

- Added a test for deeply nested types ... using the example you gave of quadruple nesting.
- Changed all object initializers so that every property assignment was on a separate line, as requested.

There were several points that I didn't make changes for, here are the rationalizations:

- I didn't change the `DescKind.Operator` output, because you can still write an `M` cref to target the member as you describe ... but because the `Operator` type is still part of the published `EcmaDesc` functionality, you can have an `EcmaDesc` with the `Operator` type, and it should be reflected in the resulting string.
- The output for explicitly implemented members continues to use the `$` character. The primary reason is that the ecma parser actually fails to parse URLs of the form used in `csc /doc` with the following error:

  > M:System.Web.SessionState.HttpSessionStateContainer#System#Web#SessionState#IHttpSessionState#Item(System.Int32)' deemed not valid`  

  ... also, the use of `$` is enshrined in both [tests](https://github.com/mono/mono/blob/b7a308f660de8174b64697a422abfc7315d07b8c/mcs/class/monodoc/Test/Monodoc.Ecma/EcmaUrlTests.cs#L176), and the output of [member nodes URLs](https://github.com/mono/mono/blob/b7a308f660de8174b64697a422abfc7315d07b8c/mcs/class/monodoc/Monodoc/providers/ecma-provider.cs#L317).
- Generics still use `<` and `>` for their type parameters. I actually started to change this, because your suggestion and reason was very good ... however, it looks like there's code that expects generics to use the angle brackets, for example [here].(https://github.com/mono/mono/blob/b7a308f660de8174b64697a422abfc7315d07b8c/mcs/class/monodoc/Monodoc/providers/EcmaDoc.cs#L519).

In each of these cases ... the changes that don't conform to the ecma spec you referenced is long-standing functionality. I propose that changing these to match the ecma spec is out of scope for this particular pr (which is just one change of several that are planned to support the `mdoc profile` functionality). A future set of work could address the output of both the `EcmaDesc.ToEcmaCref`, and `EcmaHelpSource.GetPublicUrl()` methods to bring them more inline with the ecma spec ... but I'd like to separate that out so that I can meet the short-term goals of adding the profiles.

Thoughts?

File: mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs (hunk 2 above)
Comment (jonpryor) [inline mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs]: I should have suggested this earlier, but is there a reasonable way to use `#` everywhere _except_ in URLs, and in URLs we replace `#` with `$` or some other useful ASCII character (`,`? `!`?). That would at least allow us to be consistent with established practice.

Comment (joelmartinez) [inline mcs/class/monodoc/Monodoc.Ecma/EcmaDesc.cs]: without changes to the parser/grammar, not currently ... I tested, and it fails to parse crefs that contain `#`. Much of the codebase assumes that we'll be using `$`, so it would probably be a bigger job.

**Comment by jonpryor** [comment]
From 20 days ago (sorry!):

> Do I change the existing functionality to match the ideal/spec?

You should change existing functionality to match the ideal. :-)

Very often as a "maintenance programmer," you'll want to do something but you won't have time to do what you want. Thus, the only way to do what you want is to do it as part of necessary work. ;-)

Meaning if you see an opportunity to improve code in ways you want to do *anyway*, you shouldn't pass up the opportunity, as you might not get a chance to do so again.

So unless fixing monodoc is nigh impossible, I'd always suggest fixing things to at least get *closer* to your ideal.

> The primary reason is that the ecma parser actually fails to parse URLs of the form used in csc /doc with the following error:

That sounds like a bug that should be fixed then. :-)

> ... also, the use of $ is enshrined in both existing tests, and the output of member nodes URLs.

Tests are fungible; they can be "fixed" just like anything else. `monodoc` is for internal use, has always been for internal use, and while we should strive to minimize API breaks (so we don't break existing internal tooling, such as `monodoc` the GUI browser!) I don't think we need to have a hard-and-fast "no API breakage at all" rule.

As such, would fixing these tests and the behavior they cover break anything we care about? Anything that would break Xamarin Studio or the monodoc GUI browser?

>  propose that changing these to match the ecma spec is out of scope for this particular pr (which is just one change of several that are planned to support the mdoc profile functionality).

Indeed, they should be separate PRs. It's still good to discuss the issues. :-)
